src/blocksender.cpp

```diff
@@ -84,21 +98,28 @@ void BlockSender::sendBlock(CNode& node,
 
     // We only support MSG_XTHINBLOCK, if peer wants MSG_THINBLOCK,
     // fallback to full one.
-    if (invType == MSG_BLOCK || invType == MSG_THINBLOCK)
+    if (invType == MSG_BLOCK || (invType == MSG_THINBLOCK
+                && !NodeStatePtr(node.id)->supportsCompactBlocks))
```

Comment:
I don't like this.... what if the node also supports compact blocks but is actually sending MSG_THINBLOCK? Let's not assume it's NOT MSG_THINBLOCK, especially if they have the XTHIN service bit set, which would suggest that it is.